applications/serial_lte_modem/src/slm_at_host.c

```diff
@@ -192,55 +197,50 @@ void data_send(const uint8_t *data, size_t len)
 static int uart_receive_enable(void)
 {
 	int ret;
+	int cycles = 0;
 
-	if (k_sem_take(&rx_change, K_MSEC(UART_RX_CHANGE_TIMEOUT_MS)) != 0) {
-		LOG_ERR("UART RX failed to enable");
-		(void)uart_send(FATAL_STR, sizeof(FATAL_STR) - 1);
-		return -EAGAIN;
+	if (uart_state == STATE_ENABLED) {
+		return 0;
 	}
 
+	while (uart_state == STATE_DISABLING) {
+		k_sleep(K_MSEC(10));
+		cycles++;
+		if (cycles > 10) {
+			LOG_ERR("UART RX enable failed: Timeout");
+			(void)uart_send(FATAL_STR, sizeof(FATAL_STR) - 1);
+			return -EAGAIN;
+		}
+	}
+
+	uart_state = STATE_ENABLED;
 	next_buf = uart_rx_buf[1];
 	ret = uart_rx_enable(uart_dev, uart_rx_buf[0], sizeof(uart_rx_buf[0]), UART_RX_TIMEOUT_US);
-	if (ret && ret != -EBUSY) {
+	if (ret) {
 		LOG_ERR("UART RX enable failed: %d", ret);
 		(void)uart_send(FATAL_STR, sizeof(FATAL_STR) - 1);
-		k_sem_give(&rx_change);
+		uart_state = STATE_DISABLED;
 		return ret;
 	}
 	at_buf_overflow = false;
 	at_buf_len = 0;
 
-	k_sem_give(&rx_change);
-
 	return 0;
 }
 
-static void uart_receive_disable(bool wait_for_disable)
+static void uart_receive_disable(void)
 {
 	int ret;
 
-	if (k_sem_take(&rx_change, K_MSEC(UART_RX_CHANGE_TIMEOUT_MS)) != 0) {
-		LOG_ERR("UART RX failed to disable");
-		(void)uart_send(FATAL_STR, sizeof(FATAL_STR) - 1);
+	if (uart_state == STATE_DISABLED || uart_state == STATE_DISABLING) {
 		return;
 	}
 
+	uart_state = STATE_DISABLING;
 	ret = uart_rx_disable(uart_dev);
 	if (ret) {
-		if (ret != -EFAULT) {
-			LOG_ERR("UART RX disable failed: %d", ret);
-		}
-		k_sem_give(&rx_change);
-		return;
-	}
-
-	if (wait_for_disable == true) {
-		/* Released when UART disabling completes with UART_RX_DISABLED. */
-		if (k_sem_take(&rx_change, K_MSEC(UART_RX_CHANGE_TIMEOUT_MS)) != 0) {
-			LOG_ERR("UART RX waiting for disable failed");
-			return;
-		}
-		k_sem_give(&rx_change);
+		LOG_ERR("UART RX disable failed: %d", ret);
+		uart_state = STATE_ENABLED;
```

Comment:
This is a bit tricky. Check Serial code of below.
Any disable of UART not via uart_receive_disable() would cause trouble.

	/* Signal error if RX is already enabled or if the driver is waiting
	 * for the RXTO event after a call to uart_rx_disable() to discard
	 * data from the UARTE internal RX FIFO.
	 */
	if (data->async->rx_enabled || data->async->discard_rx_fifo) {
		return -EBUSY;
	}

Reply:
Based on tests failing in CI and @junqingzou:s comments. Changing approach.